### ISSUE TITLE
TypeVar bound to Type[] of class resolves to incorrect type

### ISSUE THREAD

**Issue by d-tw** [microsoft/pyright#1586]
**Describe the bug**
Using a `T = TypeVar('T', bound=Type[SomeClass])` to indicate that `T` should be restricted to the class `SomeClass` or subclasses of `SomeClass`, it seems that the type of `T` is resolved to the _instance_ type of `SomeClass`, not the _class_ type `SomeClass`.

As an additional note, based on the screenshot below, the revealed type of `c` on line 18 is not what I'd expected, but that may be normal behaviour.

**To Reproduce**
See screenshots/code below.

**Expected behavior**
There should not be any type issues in the provided code snippet (additionally, `c` should be revealed as `MyClass`).

**Screenshots or Code**
<img width="1176" alt="image" src="https://user-images.githubusercontent.com/3317531/110242070-8dec8300-7f54-11eb-9dbd-b389f023500b.png">

```python
from typing import Type, TypeVar

class Base:
    ...

class MyClass(Base):
    ...

T = TypeVar('T', bound=Type[Base])

def register_class(cls: Type[Base]) -> None: ...

def register(cls: T) -> T:
    register_class(cls)
    return cls

c = register(MyClass)
reveal_type(c)
```

pyright output:
```sh
Loading configuration file at (...)/pyrightconfig.json
Assuming Python version 3.8
Assuming Python platform Darwin
stubPath(...)/typings is not a valid directory.
Searching for source files
Found 1 source file
(...)/sample.py
  (...)/sample.py:14:20 - error: Argument of type "T@register" cannot be assigned to parameter "cls" of type "Type[Base]" in function "register_class"
    "Base" is incompatible with "Type[Base]" (reportGeneralTypeIssues)
  (...)/sample.py:18:13 - info: Type of "c" is "c"
1 error, 0 warnings, 1 info 
Completed in 0.655sec
```

**VS Code extension or command-line**
pyright 1.1.118 / pylance v2021.3.1-pre.1

**Comment by d-tw** [comment]
Probably related, when I try a different approach, `T = TypeVar('T', bound=Base)` and using `Type[T]` in the function signature, I believe I encounter a variation of the same issue.

I would expect pyright to detect a type issue on line 16, as `cls` is of type `Type[T]`, which doesn't match the expected return type `T`.

<img width="513" alt="image" src="https://user-images.githubusercontent.com/3317531/110242295-b7f27500-7f55-11eb-9ea2-3131735c7df5.png">

```python
from typing import Type, TypeVar

class Base:
    ...

class MyClass(Base):
    ...

T = TypeVar('T', bound=Base)

def register_class(cls: Type[Base]) -> None: ...

def register(cls: Type[T]) -> T:
    register_class(cls)
    reveal_type(cls)
    return cls

c = register(MyClass)
reveal_type(c)
```

**Comment by d-tw** [comment]
One final extra detail, I'm sure one of these two functions should have a type error, given the different return types.

<img width="551" alt="image" src="https://user-images.githubusercontent.com/3317531/110242450-7f06d000-7f56-11eb-8854-f1ac3be4d9f1.png">

**Comment by erictraut** [comment]
The TypeVar should be bound to `Base`, not `Type[Base]`. If you do this and combine your two variations, it will work fine:

```python
from typing import Type, TypeVar

class Base:
    ...

class MyClass(Base):
    ...

T = TypeVar('T', bound=Base)

def register_class(cls: Type[Base]) -> None: ...

def register(cls: Type[T]) -> Type[T]:
    register_class(cls)
    return cls

c = register(MyClass)
reveal_type(c)
```

**Comment by d-tw** [comment]
Thanks for your quick reply! I didn't realise you shouldn't set a `Type[]` based bound. The code snippet you provided works, but I bump into what seems to be the originally described issue (or another part that I've misunderstood!) when I make the decorator into a parameterised decorator.


<img width="852" alt="image" src="https://user-images.githubusercontent.com/3317531/110249422-555da100-7f76-11eb-8e4c-10b6f381bc24.png">

```python
from typing import Callable, TypeVar, Type

class Base:
    ...

T = TypeVar('T', bound=Base)  # noqa: N806

def register(state_name: str, state: Type[T]): ...

def register_state(state_name: str) -> Callable[[Type[T]], Type[T]]:
    def decorator(state: Type[T]) -> Type[T]:
        register(state_name, state)
        return state

    reveal_type(decorator)

    return decorator
```

Here's the pyright output, as the error message was too wide to screenshot:

```
(...)/cls_decorator.py:15:17 - info: Type of "decorator" is "(state: Type[T@decorator]) -> Type[T@decorator]"
  (...)/cls_decorator.py:17:12 - error: Expression of type "(state: Type[T@decorator]) -> Type[T@decorator]" cannot be assigned to return type "(_p0: Type[T@register_state]) -> Type[T@register_state]"
    Type "(state: Type[T@decorator]) -> Type[T@decorator]" cannot be assigned to type "(_p0: Type[T@register_state]) -> Type[T@register_state]"
      Parameter 1: type "Type[T@register_state]" cannot be assigned to type "Type[T@decorator]"
        Type "Type[T@register_state]" cannot be assigned to type "Type[T@decorator]"
          Type "Type[T@register_state]" is incompatible with bound type "Base" for TypeVar "T@decorator"
            "Type[type]" is incompatible with "Type[Base]"
      Function return type "Type[T@decorator]" is incompatible with type "Type[T@register_state]"
        Type "Type[T@decorator]" cannot be assigned to type "Type[T@register_state]"
          Type "Type[T@decorator]" cannot be assigned to type "Type[T@register_state]" (reportGeneralTypeIssues)
```

**Comment by erictraut** [comment]
Yep, that's a bug. The constraint solver wasn't properly handling Type[T] in this edge case. A fix will be in the next release. I've used your sample above in the unit test so this doesn't regress in the future. 

https://github.com/microsoft/pyright/commit/b52f7535a13b91189eebd9be3e5265039e7c71b8

**Comment by d-tw** [comment]
Great, thanks!

**Comment by erictraut** [comment]
This is included in Pyright 1.1.119, which I just published. It will also be included in the next version of Pylance.
